RecoParticleFlow/PFProducer/src/MLPFModel.cc

```diff
@@ -0,0 +1,212 @@
+      if (!orig2->GsftrackRefPF().isNull()) {
+        charge = orig2->GsftrackRefPF()->charge();
+      }
+    } else if (type == reco::PFBlockElement::ECAL || type == reco::PFBlockElement::PS1 ||
+               type == reco::PFBlockElement::PS2 || type == reco::PFBlockElement::HCAL ||
+               type == reco::PFBlockElement::HO || type == reco::PFBlockElement::HFHAD ||
+               type == reco::PFBlockElement::HFEM) {
+      const auto& ref = ((const reco::PFBlockElementCluster*)&orig)->clusterRef();
+      if (ref.isNonnull()) {
+        eta = ref->eta();
+        phi = ref->phi();
+        px = ref->position().x();
+        py = ref->position().y();
+        pz = ref->position().z();
+        energy = ref->energy();
+        layer = ref->layer();
+        depth = ref->depth();
+      }
+    } else if (type == reco::PFBlockElement::SC) {
+      const auto& clref = ((const reco::PFBlockElementSuperCluster*)&orig)->superClusterRef();
+      if (clref.isNonnull()) {
+        eta = clref->eta();
+        phi = clref->phi();
+        px = clref->position().x();
+        py = clref->position().y();
+        pz = clref->position().z();
+        energy = clref->energy();
+      }
+    }
+
+    float typ_idx = static_cast<float>(elem_type_encoding.at(orig.type()));
+
+    //Must be the same order as in tf_model.py
+    return std::array<float, NUM_ELEMENT_FEATURES>({{typ_idx,
+                                                     pt,
+                                                     eta,
+                                                     phi,
+                                                     energy,
+                                                     layer,
+                                                     depth,
+                                                     charge,
+                                                     trajpoint,
+                                                     eta_ecal,
+                                                     phi_ecal,
+                                                     eta_hcal,
+                                                     phi_hcal,
+                                                     muon_dt_hits,
+                                                     muon_csc_hits}});
+  }
+
+  float normalize(float in) {
+    if (std::abs(in) > 1e4) {
+      return 0.0;
+    } else if (std::isnan(in)) {
+      return 0.0;
+    }
+    return in;
+  }
+
+  int arg_max(std::vector<float> const& vec) {
+    return static_cast<int>(std::distance(vec.begin(), max_element(vec.begin(), vec.end())));
+  }
+
+  reco::PFCandidate make_candidate(int pred_pid, int pred_charge, float pred_e, float pred_eta, float pred_phi) {
+    pred_phi = angle0to2pi::make0To2pi(pred_phi);
+
+    //currently, set the pT from a massless approximation.
+    //later versions of the model may predict predict both the energy and pT of the particle
+    float pred_pt = pred_e / cosh(pred_eta);
```

Comment:
for performance, consider to pass p/pt in the arguments

Reply:
Good point. We will update the model in the next iteration to predict all 4 components of the momentum vector, hence this will go away. I will keep it as it is, as the DNN inference time is likely dominating the runtime.